compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -1550,6 +1550,14 @@ object desugar {
         trees foreach collect
       case Block(Nil, expr) =>
         collect(expr)
+      case Quote(expr) =>
+        new TreeTraverser {
+          def traverse(tree: untpd.Tree)(implicit ctx: Context): Unit = tree match {
+            case Splice(expr) => collect(expr)
+            case TypSplice(expr) => collect(expr)
+            case _ => traverseChildren(tree)
+          }
+        }.traverse(expr)
```

Comment:
Why not just `collect(expr)` and handle `Splice` and `TypeSplice` as two cases in the outer match?

Reply:
Because everything directly inside of the quote is quoted code that is not part of the pattern (not in pattern mode). It only enters back in pattern mode inside the splices

For example in 
```scala
case '{ StringContext(${ExprSeq(parts2)}: _*) }
```
we use the traverser to traverse the trees `StringContext(XYZ: _*)` and then `collect` on `XYZ`, which in this case is a collect on the pattern `ExprSeq(parts2)` that will collect the variable `parts2`.